### PR TITLE
Draft

### PR DESCRIPTION
**Pull request by PatrykGala**
## Before submitting checklist

- [ ] Did you **update the CHANGELOG**? (not for test updates, internal changes/refactors or CI/CD setup)
- [ ] Did you **ask the docs owner** to review all the user-facing changes?

### CODE REVIEW

File: src/neptune/metadata_containers/safe_container.py [hunk 1]
```diff
@@ -13,34 +13,33 @@
...
-    def __init__(self, container: MetadataContainer):
-        self._container = container
 
-    def __getattr__(self, item: str) -> Any:
-        return SafeMethodWrapper(getattr(self._container, item))
+def decorator_safety(cls):
+    for name, method in inspect.getmembers(cls):
+        if (not inspect.ismethod(method) and not inspect.isfunction(method)) or inspect.isbuiltin(method):
+            continue
+        print("Decorating function %s" % name)
```
Comment (aniezurawski) [inline src/neptune/metadata_containers/safe_container.py:43]: Let's remove it or use `logger.debug()`.

File: src/neptune/metadata_containers/safe_container.py [hunk 2]
```diff
@@ -13,34 +13,33 @@
...
 
-class SafeContainer:
-    def __init__(self, container: MetadataContainer):
-        self._container = container
 
-    def __getattr__(self, item: str) -> Any:
-        return SafeMethodWrapper(getattr(self._container, item))
+def decorator_safety(cls):
+    for name, method in inspect.getmembers(cls):
+        if (not inspect.ismethod(method) and not inspect.isfunction(method)) or inspect.isbuiltin(method):
```
Comment (aniezurawski) [inline src/neptune/metadata_containers/safe_container.py:41]: `or inspect.isbuiltin(method)`

So will it decorate `__setitem__` etc. or not?

File: src/neptune/__init__.py [hunk 3]
```diff
@@ -117,36 +117,34 @@
...
 
 
-def _init_safety(obj: MetadataContainer) -> Union[MetadataContainer, SafeContainer]:
-    safety_mode = os.environ.get(NEPTUNE_SAFETY_MODE)
-    if safety_mode is None or not bool(safety_mode) or str(safety_mode).lower() == "false" or int(safety_mode) == 0:
-        return obj
-    else:
-        return SafeContainer(obj)
+def init_run(*args: Any, **kwargs: Any) -> Union[MetadataContainer, Run]:
+    @decorator_safety
```
Comment (aniezurawski) [inline src/neptune/__init__.py:131]: Can we just decorate Run, Model, etc. classes?

File: src/neptune/metadata_containers/safe_container.py [hunk 4]
```diff
@@ -13,34 +13,33 @@
...
 
+    return wrapper
 
-class SafeContainer:
-    def __init__(self, container: MetadataContainer):
-        self._container = container
 
-    def __getattr__(self, item: str) -> Any:
-        return SafeMethodWrapper(getattr(self._container, item))
+def decorator_safety(cls):
```
Comment (aniezurawski) [inline src/neptune/metadata_containers/safe_container.py:39]: It should be configurable by env variable.

**Review by aniezurawski — CHANGES_REQUESTED** [review]

File: src/neptune/metadata_containers/safe_container.py (hunk 2 above)
Comment (PatrykGala) [reply src/neptune/metadata_containers/safe_container.py:41]: List of decorated methods: 
```
Decorating function __class_getitem__
Decorating function __delitem__
Decorating function __enter__
Decorating function __exit__
Decorating function __getattr__
Decorating function __getitem__
Decorating function __init__
Decorating function __setitem__
Decorating function __subclasshook__
Decorating function _before_fork
Decorating function _define_attribute
Decorating function _fetch_entries
Decorating function _get_callback
Decorating function _get_or_create_api_object
Decorating function _get_root_handler
Decorating function _get_subpath_suggestions
Decorating function _handle_fork_in_child
Decorating function _handle_fork_in_parent
Decorating function _ipython_key_completions_
Decorating function _pop_impl
Decorating function _prepare_background_jobs
Decorating function _prepare_background_jobs_if_non_read_only
Decorating function _print_structure_impl
Decorating function _raise_if_stopped
Decorating function _shutdown_hook
Decorating function _startup
Decorating function _write_initial_attributes
Decorating function assign
Decorating function define
Decorating function exists
Decorating function fetch
Decorating function get_attribute
Decorating function get_root_object
Decorating function get_state
Decorating function get_structure
Decorating function get_url
Decorating function lock
Decorating function ping
Decorating function pop
Decorating function print_structure
Decorating function set_attribute
Decorating function start
Decorating function stop
Decorating function sync
Decorating function wait
```

**Pull request closed without merging**